File: typescript/src/resources/goals.ts
import { HttpClient } from '../http';
import { normalizePaginated, normalizeSingle } from '../normalize';
import type {
  Goal,
  CreateGoalInput,
  UpdateGoalInput,
  Task,
  Comment,
  CreateCommentInput,
  UpdateCommentInput,
  ListGoalsOptions,
  ListTasksOptions,
  ListOptions,
  PaginatedResponse,
} from '../types';

export class GoalsResource {
  constructor(private http: HttpClient) {}

  // ─── Core CRUD ───────────────────────────────────────────────────────

  async list(options?: ListGoalsOptions): Promise<PaginatedResponse<Goal>> {
    const params: Record<string, unknown> = {};
    const limit = options?.limit ?? 20;
    const offset = options?.offset ?? 0;
    if (options) {
      if (options.status) params.status = options.status;
      if (options.limit) params.limit = options.limit;
      if (options.offset) params.offset = options.offset;
    }
    // Worker returns: { goals: [...], count, total, hasMore, limit, offset }
    const raw = await this.http.get<Record<string, unknown>>('/goals', params);
    return normalizePaginated<Goal>(raw, 'goals', limit, offset);
  }

  async get(id: string): Promise<Goal> {
    // Worker returns: { goal: {...} }
    const raw = await this.http.get<Record<string, unknown>>(`/goals/${id}`);
    return normalizeSingle<Goal>(raw, 'goal');
  }

  async create(input: CreateGoalInput): Promise<Goal> {
    // Worker returns: { goal: {...} }
    const raw = await this.http.post<Record<string, unknown>>('/goals', input);
    return normalizeSingle<Goal>(raw, 'goal');
  }

  async update(id: string, input: UpdateGoalInput): Promise<Goal> {
    // Worker returns: { goal: {...} }
    const raw = await this.http.patch<Record<string, unknown>>(`/goals/${id}`, input);
    return normalizeSingle<Goal>(raw, 'goal');
  }

  async delete(id: string): Promise<void> {
    await this.http.delete(`/goals/${id}`);
  }

  // ─── Tasks under a goal ──────────────────────────────────────────────

  /** List tasks associated with a goal (requires read:tasks scope) */
  async listTasks(goalId: string, options?: ListTasksOptions): Promise<PaginatedResponse<Task>> {
    const params: Record<string, unknown> = {};
    const limit = options?.limit ?? 20;
    const offset = options?.offset ?? 0;
    if (options) {
      if (options.status) params.status = options.status;
      if (options.priority) params.priority = options.priority;
      if (options.assigned_to) params.assigned_to = options.assigned_to;
      if (options.limit) params.limit = options.limit;
      if (options.offset) params.offset = options.offset;
      if (options.search) params.search = options.search;
    }
    const raw = await this.http.get<Record<string, unknown>>(`/goals/${goalId}/tasks`, params);
    return normalizePaginated<Task>(raw, 'tasks', limit, offset);
  }

  // ─── Comments ─────────────────────────────────────────────────────────

  /** List all comments on a goal (requires read:comments scope) */
  async listComments(goalId: string, options?: ListOptions): Promise<PaginatedResponse<Comment>> {
    const params: Record<string, unknown> = {};
    const limit = options?.limit ?? 20;
    const offset = options?.offset ?? 0;
    if (options) {
      if (options.limit) params.limit = options.limit;
      if (options.offset) params.offset = options.offset;
    }
    const raw = await this.http.get<Record<string, unknown>>(`/goals/${goalId}/comments`, params);
    return normalizePaginated<Comment>(raw, 'comments', limit, offset);
  }

  /** Add a comment to a goal (requires write:comments scope) */
  async addComment(goalId: string, input: CreateCommentInput): Promise<Comment> {
    const raw = await this.http.post<Record<string, unknown>>(
      `/goals/${goalId}/comments`,
      input,
    );
    return normalizeSingle<Comment>(raw, 'comment');
  }

  /** Update a comment (requires write:comments scope) */
  async updateComment(commentId: string, input: UpdateCommentInput): Promise<Comment> {
    const raw = await this.http.patch<Record<string, unknown>>(
      `/comments/${commentId}`,
      input,
    );
    return normalizeSingle<Comment>(raw, 'comment');
  }

  /** Delete a comment (requires write:comments scope) */
  async deleteComment(commentId: string): Promise<void> {
    await this.http.delete(`/comments/${commentId}`);
  }
}

File: go/goals.go
package kairos

import "context"

// GoalsService provides goal operations.
type GoalsService struct {
	client *Client
}

// List returns a paginated list of goals.
func (s *GoalsService) List(ctx context.Context, opts *ListOptions) ([]Goal, *Pagination, error) {
	if opts == nil {
		opts = &ListOptions{}
	}

	var resp struct {
		Data       []Goal     `json:"data"`
		Pagination Pagination `json:"pagination"`
	}

	_, err := s.client.get(ctx, "/goals", opts, &resp)
	if err != nil {
		return nil, nil, err
	}

	return resp.Data, &resp.Pagination, nil
}

// Get returns a single goal by ID.
func (s *GoalsService) Get(ctx context.Context, id string) (*Goal, error) {
	var resp struct {
		Data *Goal `json:"data"`
	}

	_, err := s.client.get(ctx, "/goals/"+id, nil, &resp)
	if err != nil {
		return nil, err
	}

	return resp.Data, nil
}

// Create creates a new goal.
func (s *GoalsService) Create(ctx context.Context, input CreateGoalInput) (*Goal, error) {
	var resp struct {
		Data *Goal `json:"data"`
	}

	_, err := s.client.post(ctx, "/goals", input, &resp)
	if err != nil {
		return nil, err
	}

	return resp.Data, nil
}

// Update updates a goal.
func (s *GoalsService) Update(ctx context.Context, id string, input UpdateGoalInput) (*Goal, error) {
	var resp struct {
		Data *Goal `json:"data"`
	}

	_, err := s.client.patch(ctx, "/goals/"+id, input, &resp)
	if err != nil {
		return nil, err
	}

	return resp.Data, nil
}

// Delete deletes a goal.
func (s *GoalsService) Delete(ctx context.Context, id string) error {
	_, err := s.client.delete(ctx, "/goals/"+id, nil)
	return err
}

// ListTasks returns tasks associated with a goal (requires read:tasks scope).
func (s *GoalsService) ListTasks(ctx context.Context, goalID string, opts *ListTasksOptions) ([]Task, *Pagination, error) {
	if opts == nil {
		opts = &ListTasksOptions{}
	}

	var resp struct {
		Data       []Task     `json:"data"`
		Pagination Pagination `json:"pagination"`
	}

	_, err := s.client.get(ctx, "/goals/"+goalID+"/tasks", opts, &resp)
	if err != nil {
		return nil, nil, err
	}

	return resp.Data, &resp.Pagination, nil
}

// ListComments returns comments on a goal (requires read:comments scope).
func (s *GoalsService) ListComments(ctx context.Context, goalID string, opts *ListOptions) ([]Comment, *Pagination, error) {
	if opts == nil {
		opts = &ListOptions{}
	}

	var resp struct {
		Data       []Comment  `json:"data"`
		Pagination Pagination `json:"pagination"`
	}

	_, err := s.client.get(ctx, "/goals/"+goalID+"/comments", opts, &resp)
	if err != nil {
		return nil, nil, err
	}

	return resp.Data, &resp.Pagination, nil
}

// AddComment adds a comment to a goal (requires write:comments scope).
func (s *GoalsService) AddComment(ctx context.Context, goalID string, input CreateCommentInput) (*Comment, error) {
	var resp struct {
		Data *Comment `json:"data"`
	}

	_, err := s.client.post(ctx, "/goals/"+goalID+"/comments", input, &resp)
	if err != nil {
		return nil, err
	}

	return resp.Data, nil
}

// UpdateComment updates a comment (requires write:comments scope).
func (s *GoalsService) UpdateComment(ctx context.Context, commentID string, input UpdateCommentInput) (*Comment, error) {
	var resp struct {
		Data *Comment `json:"data"`
	}

	_, err := s.client.patch(ctx, "/comments/"+commentID, input, &resp)
	if err != nil {
		return nil, err
	}

	return resp.Data, nil
}

// DeleteComment deletes a comment (requires write:comments scope).
func (s *GoalsService) DeleteComment(ctx context.Context, commentID string) error {
	_, err := s.client.delete(ctx, "/comments/"+commentID, nil)
	return err
}
